source/k64f_eth_nanostack_port.c

```diff
@@ -501,236 +361,275 @@ static int8_t arm_eth_phy_k64f_address_write(phy_address_type_e address_type, ui
-    mac_config_ptr->duplex = kEnetCfgFullDuplex;
-    mac_config_ptr->macCtlConfigure = kEnetRxCrcFwdEnable | kEnetRxFlowControlEnable,
-    mac_config_ptr->isTxAccelEnabled = true;
-    mac_config_ptr->isRxAccelEnabled = true;
-    mac_config_ptr->isStoreAndFwEnabled = false;
-    rxAcceler.isIpcheckEnabled = false;
-    rxAcceler.isMacCheckEnabled = true;
-    rxAcceler.isPadRemoveEnabled = false;
-    rxAcceler.isProtocolCheckEnabled = false;
-    rxAcceler.isShift16Enabled = false;
-    mac_config_ptr->rxAcceler = rxAcceler;
-    txAcceler.isIpCheckEnabled = false;
-    txAcceler.isProtocolCheckEnabled = false;
-    txAcceler.isShift16Enabled = false;
-    mac_config_ptr->txAcceler = txAcceler;
-    mac_config_ptr->isVlanEnabled = false;
-    mac_config_ptr->isPhyAutoDiscover = true;
-    mac_config_ptr->miiClock = ENET_MII_CLOCK;
-
-    ethernet_iface_ptr->macContextPtr = MEM_ALLOC (sizeof(enet_mac_context_t));
-    if (!ethernet_iface_ptr->macContextPtr) {
-        tr_debug("Memory Allocation Failed. mac_context_ptr");
-        return -1;
-    }
-    memset(ethernet_iface_ptr->macContextPtr, 0, sizeof(enet_mac_context_t));
-    ethernet_iface_ptr->maxFrameSize = ENET_ETH_MAX_FLEN;
-    ethernet_iface_ptr->macCfgPtr = mac_config_ptr;
-
-    /* Setup PHY related configuration settings*/
-    ethernet_iface_ptr->phyCfgPtr = &enetPhyCfg[BOARD_DEBUG_ENET_INSTANCE];
-    ethernet_iface_ptr->macApiPtr = &g_enetMacApi;
-    ethernet_iface_ptr->phyApiPtr = (void *)&g_enetPhyApi;
-
-    /* Setup RX buffer descriptors*/
-    retval = k64f_eth_rx_buf_desc_setup(buf_desc_ring, &rxbdCfg);
-    if (retval==-1){
-        return retval;
-    }
-
-    /* Setup TX buffer descriptors*/
-    retval = k64f_eth_tx_buf_desc_setup(buf_desc_ring, &txbdCfg);
-    if (retval==-1){
-        return retval;
-    }
-
-    /* Initialize MAC layer*/
-    if (enet_mac_init(ethernet_iface_ptr, &rxbdCfg, &txbdCfg)
-            == kStatus_ENET_Success) {
-        /* Initialize PHY layer*/
-        if (ethernet_iface_ptr->macCfgPtr->isPhyAutoDiscover) {
-            if (((enet_phy_api_t *) (ethernet_iface_ptr->phyApiPtr))->phy_auto_discover(
-                    ethernet_iface_ptr) != kStatus_PHY_Success) {
-                tr_debug("INIT_MAC. Auto discover fail.");
-                return -1;
-            }
+    bool link = false;
+
+    eth_if_lock();
+    PHY_GetLinkStatus(ENET, 0, &link);
+    if (link != link_currently_up) {
+        link_currently_up = link;
+        if (link) {
+            phy_duplex_t phy_duplex;
+            phy_speed_t phy_speed;
+            PHY_GetLinkSpeedDuplex(ENET, 0, &phy_speed, &phy_duplex);
+            /* Poke the registers*/
+            ENET_SetMII(ENET, (enet_mii_speed_t)phy_speed, (enet_mii_duplex_t)phy_duplex);
+            ENET_EnableInterrupts(ENET, kENET_RxFrameInterrupt|kENET_TxFrameInterrupt);
+        } else {
+            ENET_DisableInterrupts(ENET, kENET_RxFrameInterrupt|kENET_TxFrameInterrupt);
```

Comment:
I'm a bit nervous about transmitted frames getting stuck in the system if the link is down. We certainly don't have any code to stop people trying to transmit if the link is down, which means we presumably queue them up, and presumably they time out in the hardware, and we never reclaim the buffer (until the cable is plugged back in).

Even if we did block new transmissions, there could presumably be stuff lingering in the system at the instant of link loss.

I'm inclined to always leave these interrupts enabled, regardless of link state. Can you see any reason not to do so?